# Hacker news data set
* hacker news is a popular site where technology related stories and posts are voted and commented upon similar to reddit.hacker newa is extermely popular in technology and start up circles, and the posts that make it to the top of hacker listings can get hundreds of thousands of visitors as a result.
* In this project we will work with a data set of subbmissions to popular techology site [hacker news.](https://news.ycombinator.com/)
* You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.
* We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. and users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. 

# introduction:
* first we will open the hacker_news.csv file and make it a list of lists and remove the header row from it.

In [3]:
from csv import reader
opened_file=open("hacker_news.csv")
read_file=reader(opened_file)
hn=list(read_file)
header=hn[0]
hn=hn[1:]

In [4]:
print(header)
print('\n')
for i in hn[:5]:
    print(i)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']
['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']
['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']
['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']
['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']


## seperating the data set into 3 parts:
* the data set which we are given is divided into three parts :
1. ask: it stores all the post titles beginning with ASK HN.
2. show: it stores all the post titles beginning with SHOW HN.
3. other: it stores other files which dont have either ASK HN or SHOW HN in their titles.

In [5]:
ask=[]
show=[]
other=[]
for i in hn:
    title=i[1]
    title_l=title.lower()
    if title_l.startswith("ask hn")==True:
        ask.append(i)
    elif title_l.startswith("show hn")==True:
        show.append(i)
    else:
        other.append(i)
print(len(ask))
print(len(show))
print(len(other))

1744
1162
17194


* we will now print the first five rows of ask list and you can observe that all the titles in this list start with ASK HN.
* we will also going to print the first five rows of show list and all the titles start with SHOW HN.

In [6]:
for i in ask[:5]:
    print(i)
print('')
print('')
for j in show[:5]:
    print(j)

['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']
['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']
['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']
['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']
['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']


['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03']
['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']
['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', 

## determining the average number of comments in ask and show lists .
* to find the average number of comments in ask and show lists we need to first initiate total_ask_c and total_show_c to 0
* loop through the ask list, number of comments is in the fifth coloumn in ask list we need to get the element at index 4 in each row. dont forget to convert the value to integer type to perform addition to it.
* calculate comments and assign it to variable total_ask_c .
* similarly do the same operations on show list and find the total_show_c
* after that calculate average comments for both the lists.

In [7]:
total_ask_c=0
total_show_c=0
for i in ask:
    comments_a=int(i[4])
    total_ask_c+=comments_a
for i in show:
    comments_s=int(i[4])
    total_show_c+=comments_s
print("total ask comments:  ",total_ask_c)
print("total show comments:  ",total_show_c)
average_ask_c=total_ask_c/len(ask)
average_show_c=total_show_c/len(show)
print("average ask comments:  ",average_ask_c)
print("average show comments: ",average_show_c)

total ask comments:   24483
total show comments:   11988
average ask comments:   14.038417431192661
average show comments:  10.31669535283993


## function method to find the average comments: 

In [8]:
def average(lis,index):
    total_c=0
    for i in lis:
        comments=int(i[index])
        total_c+=comments
        average_c=total_c/len(lis)
    print("total comments:  ",total_c)
    print("average comments:  ",average_c)
    

In [9]:
average(ask,4)

total comments:   24483
average comments:   14.038417431192661


In [10]:
average(ask,4)

total comments:   24483
average comments:   14.038417431192661


* now we will create 2 dictionaries which will store counts and comments per hour
* first import datetime module as dt because we will need it to convert the created_at coloumn at index 6 in hour.
* then create an empty list named result list.which we will be a list of lists
* append the date and comments to the list. here date and comments form for each row form a list and stored into the bigger list result_list
* create 2 dictionaries counts_by_hour and comments_by_hour which will be a frequency table where the key is hour and the value is number of posts , number of comments for that hour.

In [11]:
import datetime as dt
result_list=[]
for i in ask:
    result_list.append([i[6],int(i[4])])#i[6]=created_at , i[4]=comments
counts_by_hour={}
comments_by_hour={}
date_format = "%m/%d/%Y %H:%M"
for i in result_list:
    date=i[0]
    comment=i[1]
    hour=dt.datetime.strptime(date, date_format).strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=comment
    else:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=comment
print(counts_by_hour)
print(comments_by_hour)

{'19': 110, '21': 109, '01': 60, '11': 58, '13': 85, '06': 44, '04': 47, '14': 107, '10': 59, '03': 54, '07': 34, '18': 109, '17': 100, '12': 73, '08': 48, '20': 80, '05': 46, '16': 108, '15': 116, '09': 45, '00': 55, '02': 58, '22': 71, '23': 68}
{'19': 1188, '21': 1745, '01': 683, '11': 641, '13': 1253, '06': 397, '04': 337, '14': 1416, '10': 793, '03': 421, '07': 267, '18': 1439, '17': 1146, '12': 687, '08': 492, '20': 1722, '05': 464, '16': 1814, '15': 4477, '09': 251, '00': 447, '02': 1381, '22': 479, '23': 543}


* we will now create a list which store lists within it and each sublist is of 2 elements where the first element is hour and the second element is average comments for that hour

In [12]:
avg_by_hour=[]
for hr in comments_by_hour:
    avg_by_hour.append([hr,comments_by_hour[hr]/counts_by_hour[hr]])
print(avg_by_hour)

[['19', 10.8], ['21', 16.009174311926607], ['01', 11.383333333333333], ['11', 11.051724137931034], ['13', 14.741176470588234], ['06', 9.022727272727273], ['04', 7.170212765957447], ['14', 13.233644859813085], ['10', 13.440677966101696], ['03', 7.796296296296297], ['07', 7.852941176470588], ['18', 13.20183486238532], ['17', 11.46], ['12', 9.41095890410959], ['08', 10.25], ['20', 21.525], ['05', 10.08695652173913], ['16', 16.796296296296298], ['15', 38.5948275862069], ['09', 5.5777777777777775], ['00', 8.127272727272727], ['02', 23.810344827586206], ['22', 6.746478873239437], ['23', 7.985294117647059]]


# Sorting and printing values from a list of lists

In [13]:
swap_avg_by_hour=[]
for i in avg_by_hour:
    swap_avg_by_hour.append([i[1],i[0]])
print(swap_avg_by_hour)
sorted_swap=sorted(swap_avg_by_hour,reverse=True)
print("")
print("")
print("sorted order: ")
print(sorted_swap)

[[10.8, '19'], [16.009174311926607, '21'], [11.383333333333333, '01'], [11.051724137931034, '11'], [14.741176470588234, '13'], [9.022727272727273, '06'], [7.170212765957447, '04'], [13.233644859813085, '14'], [13.440677966101696, '10'], [7.796296296296297, '03'], [7.852941176470588, '07'], [13.20183486238532, '18'], [11.46, '17'], [9.41095890410959, '12'], [10.25, '08'], [21.525, '20'], [10.08695652173913, '05'], [16.796296296296298, '16'], [38.5948275862069, '15'], [5.5777777777777775, '09'], [8.127272727272727, '00'], [23.810344827586206, '02'], [6.746478873239437, '22'], [7.985294117647059, '23']]


sorted order: 
[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.4109589

In [15]:
for i in sorted_swap[:5]:
    print("{}:00 : {:.2f} average comments per post.".format(i[1],i[0]))
    

15:00 : 38.59 average comments per post.
02:00 : 23.81 average comments per post.
20:00 : 21.52 average comments per post.
16:00 : 16.80 average comments per post.
21:00 : 16.01 average comments per post.


* The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

* According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# conclusions 
* in this project our aim was to find which of the two ask posts or show posts more likely received comments and we found out that ask posts received more comments as the average comments for ask posts was 14.03 compared to average show comments was 10.31
* Then we had to find the hour of the day for which ask posts had maximum average comments and it was the 15th hour of the day.The timezone used is eastetrn time in the US. So, we could write 15:00 as 3:00 pm est. which converted to time in india is 12:30 am in the morning which looks like an inapporiate time for maximum comments for ask posts in india.